### PR TITLE
Merging ifr-gains into master, finally

### PR DESCRIPTION
**Pull request by o-smirnov**
This contains fixes to many issues (yes, I should have been doing smaller feature branches. I'll be better in the future). 

@SpheMakh @IanHeywood, would you please run your stuff against the ``ifr-gains-merge-master`` branch to make sure the merge hasn't broken anything? Please report feedback here so @JSKenyon can merge.

### CODE REVIEW

**Comment by IanHeywood** [comment]
First test failed with the following:

```
(cubienv) heywoodl@aslxdaq04:/data/disk1/meerkat/trouble$ gocubical cube_phasecal.parset 
 - 10:15:44 - shm_utils          | reaping 1 shared memory objects associated with 1 dead cubical processes
 - 10:15:44 - cc                 [0.1/0.1 0.8/0.8 0.1Gb] set global console verbosity level 1
 - 10:15:44 - main               [0.1/0.1 0.8/0.8 0.1Gb] Exiting with exception: TypeError(expected string or buffer)
 Traceback (most recent call last):
  File "/home/heywoodl/Software/CubiCal/cubical/main.py", line 192, in main
    logger.setGlobalLogVerbosity(GD["log"]["file-verbose"])
  File "/home/heywoodl/Software/CubiCal/cubical/tools/logger.py", line 273, in setGlobalLogVerbosity
    if type(element) is int or re.match("^[0-9]+$", element):
  File "/home/heywoodl/venv/cubienv/lib/python2.7/re.py", line 141, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or buffer
```

I was just trying a simple phasecal but it looks like it doesn't like the verbosity settings. Parset is below, which I derived by running it with no inputs and then editing the generated cubical.parset.

```
[data]
ms = 1505632500_wtspec_NGC4993.ms
column = DATA 
time-chunk = 90
freq-chunk = 64 
chunk-by = SCAN_NUMBER 
chunk-by-jump = 1.0 
single-chunk =  

[sel]
field = 0 
ddid = None 
taql =  
chan =  

[model]
list = MODEL_DATA 
ddes = False 
beam-pattern = None 
beam-l-axis = None 
beam-m-axis = None 

[montblanc]
device-type = CPU 
dtype = float 
feed-type = linear 
mem-budget = 4096 
verbosity = WARNING 

[weight]
column = WEIGHT_SPECTRUM 

[flags]
apply = legacy 
auto-init = legacy 
save = cubical 
reinit-bitflags = False 
tf-chisq-median = 1.2 
tf-np-median = 0.5 
time-density = 0.5 
chan-density = 0.5 
ddid-density = 0.5 

[sol]
jones = G 
precision = 32 
delta-g = 1e-06 
delta-chi = 1e-06 
chi-int = 5 
last-rites = True 
stall-quorum = 0.99 
min-bl = 200.0 
max-bl = 0.0 
subset =  

[bbc]
load-from =  
apply-2x2 = False 
save-to = {data[ms]}/BBC-field:{sel[field]} 
per-chan = True 

[dist]
ncpu = 5 
min-chunks = 32 

[out]
name = cubical 
mode = sc 
column = CORRECTED_DATA 
plots = True 
plots-show = True 

[log]
memory = True 
boring = False 
append = False 
verbose = 1
file-verbose = None 

[debug]
pdb = False 

[misc]
random-seed = None 
parset-version = 0.1 

[j1]
solvable = True 
label = G 
type = phase-diag 
load-from =  
save-to = {data[ms]}/G-field:{sel[field]} 
dd-term = False 
update-type = full 
time-int = 3 
freq-int = 6 
clip-low = 0.4 
clip-high = 1.6 
clip-after = 5 
max-iter = 20
conv-quorum = 0.99 
term-iters = 10 
ref-ant = None 

[j2]
solvable = True 
label = B 
type = complex-2x2 
load-from =  
save-to = {data[ms]}/B-field:{sel[field]} 
dd-term = False 
update-type = full 
time-int = 1 
freq-int = 1 
clip-low = 0.1 
clip-high = 10.0 
clip-after = 5 
max-iter = 200 
conv-quorum = 0.99 
term-iters = 2 
ref-ant = None 

[j3]
solvable = True 
label = dE 
type = complex-2x2 
load-from =  
save-to = {data[ms]}/dE-field:{sel[field]} 
dd-term = False 
update-type = full 
time-int = 1 
freq-int = 1 
clip-low = 0.1 
clip-high = 10.0 
clip-after = 5 
max-iter = 200 
conv-quorum = 0.99 
term-iters = 2 
ref-ant = None 
```

**Comment by o-smirnov** [comment]
Ahh, silly error. Can you pull and try again please?

**Comment by IanHeywood** [comment]
Thanks, it's running now. I'll let you know how it goes.

**Comment by IanHeywood** [comment]
```
Time taken for solve and correct: 291.837603092 seconds
```

**Comment by o-smirnov** [comment]
Sounds better than 6000 seconds!

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by IanHeywood** [comment]
You'll see my parset has:

```
min-bl = 200.0 
max-bl = 0.0 
```

which seems to do some horrible things to the data. I'm guessing the solver ignores these settings if they're both zero, but takes them both into account if one of them isn't. 

Explicitly setting max-bl to something huge seems to get around it, but the defaults imply that 0.0 means no selection. Maybe the default for max-bl should be something appropriately large.

**Comment by IanHeywood** [comment]
These look sane though, so maybe something else is up...

```
11:13:16 - data_handler       [P1] [3.2/3.5 4.8/5.1 3.2Gb] applying solvable baseline cutoff deselects 1634 rows
```

**Comment by IanHeywood** [comment]
Spectral window selection (via --sel-ddid) on VLA data doesn't seem to work. Logs are available [here](http://www-astro.physics.ox.ac.uk/~ianh/meerkat/cubical/spw_fail/).

Note the different behaviour for SPW0 and the rest.

**Comment by o-smirnov** [comment]
For the first thing, strange, do you see different behaviour with max-bl 0 and max-bl very large? Because the logic looks right to me: https://github.com/ratt-ru/CubiCal/blob/ifr-gains-merge-master/cubical/data_handler.py#L436

That said, I spotted a problem with ``--sol-subset`` for which I'll push a fix shortly, but you're not using that...

I also see two problems with the DDID selection code. Are you running this on a Rhodes machine where I can test quickly?

**Comment by IanHeywood** [comment]
I've gone and deleted my intermediate images but the different behaviour was very large negative bowls in the --min-bl=200, --max-bl=0 case, which went away when I set --max-bl=1000000. I'll repeat the test to make sure I didn't change anything else, and get back to you with pictures.

I'm running all this on my local desktop machine. I'm happy to quickly run tests on your behalf if you want.

Cheers.

**Comment by o-smirnov** [comment]
Can you compare logs for the two cases? In particular, the line you spotted that says "applying solvable baseline cutoff deselects X rows"?

Have you got a multiple-DDID VLA MS somewhere on a Rhodes machine that I can quickly access?

**Comment by IanHeywood** [comment]
Yep, will have a look at the logs for the bl selection.

There's a VLA MS here:
```
/vault/ianh/14A-012/MACSJ0416/S-BAND/C/sb29729009.eb29973947_MACSJ0416_wtspec.ms
```
which should have SPWs in, although I can't vouch for the quality of whatever is in the data columns. (I can clear this lot out from vault actually, as it's all been superseded, let me know when you're done.)

**Comment by o-smirnov** [comment]
OK, so I pushed a partial fix. Selecting a starting DDID>0 did not work. This is now fixed. Outstanding issues are:

* The mysterious ``--sol-max-bl`` behaviour -- waiting for more feedback from @IanHeywood 

* Plotting BBCs on @IanHeywood's MS fails. Workaround is to run with ``--bbc-save-to None`` to disable this for now. I'd like a live copy of this MS somewhere to test the problem properly.

**Comment by IanHeywood** [comment]
MS in question is here:

```
/net/tina/vault-tina/ianh/17A-170/sb33832967.eb33979842_VIDEO_VLA1_C.ms
```

**Comment by o-smirnov** [comment]
Cool. Did you try disabling BBCs?

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by o-smirnov** [comment]
Oh and upload the parsets as well please.

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by o-smirnov** [comment]
@IanHeywood reminder to send the parsets...

**Comment by IanHeywood** [comment]
Thanks, I've uploaded them [here](http://www-astro.physics.ox.ac.uk/~ianh/meerkat/cubical/spw_fail/).

**Comment by o-smirnov** [comment]
OK, pretty sure @IanHeywood's breakage is due to #88, which I'll fix shortly. The remaining problem, with ``--max-bl``, is still to be verified.

**Comment by o-smirnov** [comment]
OK @IanHeywood try again please. You've got a weird edge case there: at least one baseline has just one unflagged channel. This made the solution interpolation code (which is invoked for the post-mortem plots) cry. I *think* I made it robust against this sort of thing now, but more testing is needed of course.

**Comment by IanHeywood** [comment]
I get this one on each loop now:

```
 - 19:54:26 - data_handler       [P1] [0.2/1.2 1.1/2.3 0.7Gb] saving CORRECTED_DATA for MS rows 0~89
699
 - 19:54:26 - main               [P1] [0.2/1.2 1.1/2.3 0.7Gb] I/O handler for load None save -1 fail
ed with exception: operands could not be broadcast together with shapes (0,26,26,2,2) (64,26,26,2,2)
 (0,26,26,2,2)
 - 19:54:26 - main               [P1] [0.2/1.2 1.1/2.3 0.7Gb] Traceback (most recent call last):
  File "/home/heywoodl/Software/CubiCal/cubical/main.py", line 453, in _io_handler
    solver.ifrgain_machine.accumulate(sd)
  File "/home/heywoodl/Software/CubiCal/cubical/machines/ifr_gain_machine.py", line 150, in accumulate
    self._ddh_sum[subdict['ifrgains:freqslice__'],...] += ddh
  File "/home/heywoodl/.local/lib/python2.7/site-packages/numpy/ma/core.py", line 4169, in __iadd__
    self._mask += m
ValueError: operands could not be broadcast together with shapes (0,26,26,2,2) (64,26,26,2,2) (0,26,26,2,2)
```

**Comment by IanHeywood** [comment]
Logs are in the usual, place: *2017-11-06-19-51-56.log

**Comment by o-smirnov** [comment]
OK, that was an issue with spw selection. Now fixed -- @IanHeywood please pull and retry.

**Comment by IanHeywood** [comment]
Thanks, looks to be working now. Takes about 10 seconds to do the pre-ifr-gains part per SPW.

**Comment by IanHeywood** [comment]
> The remaining problem, with --max-bl, is still to be verified.

Verified as being user error, apologies.

**Comment by IanHeywood** [comment]
Recent tweaks to data_handler seem to break the use of montblanc?

```
 - 10:14:28 - data_handler       [P1] [0.1/0.1 1.1/1.1 0.1Gb] reading tile for MS rows 0~35759
 - 10:14:29 - main               [P1] [1.2/1.2 5.5/5.5 0.6Gb] I/O handler for load 0 save None failed with exception: 'Tile' object has no attribute 'nfreq'
 - 10:14:29 - main               [P1] [1.2/1.2 5.5/5.5 0.6Gb] Traceback (most recent call last):
  File "/home/heywoodl/Software/CubiCal/cubical/main.py", line 459, in _io_handler
    Tile.tile_list[load].load(load_model=load_model)
  File "/home/heywoodl/Software/CubiCal/cubical/data_handler.py", line 339, in load
    expected_nrows, sort_ind, row_identifiers = self.prep_for_montblanc()
  File "/home/heywoodl/Software/CubiCal/cubical/data_handler.py", line 508, in prep_for_montblanc
    expected_nrows = n_bl*ntime*len(self.nfreq)
AttributeError: 'Tile' object has no attribute 'nfreq'
 - 10:14:29 - main               [0.2/0.2 1.1/1.1 0.6Gb] Exiting with exception: AttributeError('Tile' object has no attribute 'nfreq')
 Traceback (most recent call last):
  File "/home/heywoodl/Software/CubiCal/cubical/main.py", line 349, in main
    if not done or not io_futures[itile].result():
  File "/home/heywoodl/venv/cubienv/local/lib/python2.7/site-packages/concurrent/futures/_base.py", line 422, in result
    return self.__get_result()
  File "/home/heywoodl/venv/cubienv/local/lib/python2.7/site-packages/concurrent/futures/_base.py", line 381, in __get_result
    raise exception_type, self._exception, self._traceback
AttributeError: 'Tile' object has no attribute 'nfreq'
```

**Comment by o-smirnov** [comment]
Bah, that's my scattershot approach to bugfixing biting you. That's what I get for trying to fix things in a rush while on the road. OK, now *really* fixed. I think...

**Comment by IanHeywood** [comment]
Off it goes! Cheers.

**Comment by o-smirnov** [comment]
OK all the bells and features on this branch seem to work with NGC4993. @IanHeywood, @SpheMakh, final test drives / feedback before we merge it in please?

**Comment by IanHeywood** [comment]
I think I've screwed up here, I generated a new ssh key, and although I've replaced my old ssh key in the Github settings I no longer have permission to update this branch. HALP PLZ.

```
heywoodl@aslxdaq04:~/Software/CubiCal$ git pull
sign_and_send_pubkey: signing failed: agent refused operation
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by o-smirnov** [comment]
Re-upload the key? *not sure*

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by IanHeywood** [comment]
I changed the URL in the .git/config file to the https one, and that seems to have worked. Not sure what I can do beyond re-upload my ssh key, which I already did.

**Comment by o-smirnov** [comment]
Yep, https will always work for reading. Won't let you commit easily, but it's not like you're about to commit anything anyway. We can sort out ssh next week.

**Comment by SpheMakh** [comment]
add `cubical.plots` to package data. This wouldn't raise an *ImportError* exception if installed in dev (-e) mode but did in standard installation 
c8df656

**Comment by o-smirnov** [comment]
I'm happy with this one. @SpheMakh hasn't complained. Can we merge and make a new release?

**Comment by o-smirnov** [comment]
@JSKenyon maybe you can review?

**Comment by JSKenyon** [comment]
@o-smirnov Will look over it during the conference. Sure it will be ok, but I think I should familiarise myself with it regardless.

File: cubical/data_handler.py [hunk 1]
```diff
@@ -1119,6 +1287,34 @@ def fetch(self, *args, **kwargs):
 
         return self.data.getcol(*args, **kwargs)
 
+    def fetchslice(self, column, startrow, nrows):
+        """
+        Like fetch, but assumes a column of NFREQxNCORR shape and applies the channel slice
+        """
```
Comment (JSKenyon) [inline cubical/data_handler.py]: Document fully please.

File: cubical/data_handler.py [hunk 2]
```diff
@@ -1119,6 +1287,34 @@ def fetch(self, *args, **kwargs):
...
+        if self._channel_slice == slice(None):
+            return self.data.getcol(column, startrow, nrows)
+        return self.data.getcolslice(column, self._ms_blc, self._ms_trc, [], startrow, nrows)
+
+    def putslice(self, column, value, startrow, nrows):
+        """
+        The put equivalent of fetchslice
+        """
```
Comment (JSKenyon) [inline cubical/data_handler.py]: Document fully please.

File: cubical/machines/abstract_machine.py [hunk 3]
```diff
@@ -361,18 +361,20 @@ def __init__(self, machine_cls, grid, double_precision, apply_only, global_optio
             self.init_solutions()
 
         def init_solutions(self):
-            """Internal method. Initializes solution databases. Note that this is reimplemented in JonesChain."""
-            self._init_solutions(self.jones_label, self._make_filename(self.jones_options["load-from"]),
-                                 self.solvable and self._make_filename(self.jones_options["save-to"]),
+            """Internal method, called from constructor. Initializes solution databases.
+            Default behaviour uses the _init_solutions() implementation below.
+            Note that this is reimplemented in JonesChain to collect solution info from chain."""
```
Comment (JSKenyon) [inline cubical/machines/abstract_machine.py]: If the docstring is not single-line, please use the standard, the quotes should appear before on separate lines.

File: cubical/machines/abstract_machine.py [hunk 4]
```diff
@@ -387,21 +389,38 @@ def _init_solutions(self, label, load_from, save_to, exportables):
...
-                    else:
-                        dtype = type(empty_value)
-                    db.define_param("{}:{}".format(label, sol_label), dtype, axes, empty=empty_value,
-                                    interpolation_axes=["time", "freq"], grid=self.grid)
+                    self.define_param(save_to, "{}:{}".format(label, sol_label), empty_value, axes)
                 print>> log(0), "{} solutions will be saved to {}".format(label, save_to)
 
+        def define_param(self, save_to, name, empty_value, axes,
+                          interpolation_axes=("time", "freq")):
+            """Helper method for _init_solutions(). Defines a parameter to be saved"""
```
Comment (JSKenyon) [inline cubical/machines/abstract_machine.py]: Please document fully.

File: cubical/machines/abstract_machine.py [hunk 5]
```diff
@@ -418,19 +437,19 @@ def export_solutions(self, gm, subdict):
                 subdict["{}:grid__".format(name)]  = grid
                 subdict["{}:flags__".format(name)] = value.mask
 
+        def get_solution_db(self, name):
+            return self._save_sols[name]
```
Comment (JSKenyon) [inline cubical/machines/abstract_machine.py]: Please add single-line docsting.

File: cubical/machines/abstract_machine.py [hunk 6]
```diff
@@ -418,19 +437,19 @@ def export_solutions(self, gm, subdict):
...
+
         def save_solutions(self, subdict):
             """Saves a slice of the solutions from a dictionary to the database.
-            This is called in an I/O process"""
+            This is called in the I/O process"""
```
Comment (JSKenyon) [inline cubical/machines/abstract_machine.py]: Add arguments and type to docstring.

File: cubical/plots/ifrgains.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+matplotlib.use("Agg")
+
+import numpy as np
+from cubical.tools import logger
+log = logger.getLogger("plots")
+
+from cubical.plots import DPI, ZOOM
+
+
+def _cmp_antenna (sa,sb):
```
Comment (JSKenyon) [inline cubical/plots/ifrgains.py]: This looks like 2 space indentation. Kill it with fire.

File: cubical/solver.py [hunk 8]
```diff
@@ -8,16 +8,22 @@
 import numpy as np
 import traceback
 from cubical.tools import logger, ModColor
-from cubical.data_handler import FL, Tile
+from cubical.data_handler import Tile
+from cubical.flagging import FL
 from cubical.machines import jones_chain_machine
 from cubical.statistics import SolverStats
+from pdb import set_trace as BREAK  # useful: can set static breakpoints by putting BREAK() in the code
```
Comment (JSKenyon) [inline cubical/solver.py]: This should probably be removed once debugging is finished.

File: cubical/MBTiggerSim.py [hunk 9]
```diff
@@ -120,7 +119,11 @@ def uvw(self, context):
         for ti, t in enumerate(xrange(t_low, t_high)):
             # Inspection confirms that this achieves the same effect as
             # ant_uvw[ti,1:na,:] = ...getcol(UVW, ...).reshape(na-1, -1)
-            ant_uvw[ti,1:na,:] = self._uvwco[self.sort_ind, ...][t*nbl:t*nbl+na-1, :]
+            try:
+                ant_uvw[ti,1:na,:] = self._uvwco[self.sort_ind, ...][t*nbl:t*nbl+na-1, :]
+            except:
+                print ti,self.sort_ind,max(self.sort_ind),t,nbl,na
+                raise
```
[the remark this answers is not in the document]
Comment (o-smirnov) [reply cubical/MBTiggerSim.py]: Leftovers from debugging. WIll remove.

File: cubical/TiggerSourceProvider.py [hunk 10]
```diff
@@ -18,25 +18,31 @@
...
+            
         """
 
-        self._tile = tile
-        self._handler = tile.handler
-        self._sm = Tigger.load(self._handler.sm_name)
-        self._phase_center = self._handler._phadir
-        self._use_ddes = self._handler.use_ddes
+#        self._tile = tile
+#        self._handler = tile.handler
```
[the remark this answers is not in the document]
Comment (o-smirnov) [reply cubical/TiggerSourceProvider.py]: Yep.

File: cubical/data_handler.py [hunk 11]
```diff
@@ -18,17 +21,42 @@
...
-    from montblanc.impl.rime.tensorflow.sources import CachedSourceProvider, FitsBeamSourceProvider
 
 from cubical.tools import logger, ModColor
-log = logger.getLogger("data_handler")
+log = logger.getLogger("data_handler", 1)
+
+def _parse_slice(arg, what="slice"):
+    """Helper function. Parses an string argument into a slice. 
+    Supports e.g. "5~7" (inclusive range), "5:8" (pythonic range)
+    """
```
[the remark this answers is not in the document]
Comment (o-smirnov) [reply cubical/data_handler.py]: Done.

File: cubical/data_handler.py [hunk 12]
```diff
@@ -278,85 +299,107 @@ def load(self, load_model=True):
...
+                # populate directions of this model
+                for idir,dirname in enumerate(self.handler.model_directions):
+                    if dirname in dirmodels:
+                        # loop over additive components
+                        for model_source, cluster in dirmodels[dirname]:
+                            # see if data for this model is already loaded
+                            if model_source in loaded_models:
+                                print>>log(1),"  reusing {}{} for model {} direction {}".format(model_source,
+                                                "" if not cluster else ("()" if cluster == 'die' else "({})".format(cluster)),
+                                                imod, idir)
```
[the remark this answers is not in the document]
Comment (o-smirnov) [reply cubical/data_handler.py]: Done.

File: cubical/data_handler.py [hunk 13]
```diff
@@ -278,85 +299,107 @@ def load(self, load_model=True):
...
+                                model = loaded_models[model_source][cluster]
+                                print>> log(1), "  using {}{} for model {} direction {}".format(model_source,
+                                                  "" if not cluster else
+                                                        ("()" if cluster == 'die' else "({})".format(cluster)),
+                                                  imod, idir)
+
+                            # finally, add model in at correct slot
+                            movis[idir, imod, ...] += model
+
+            del loaded_models
```
[the remark this answers is not in the document]
Comment (o-smirnov) [reply cubical/data_handler.py]: Just yielding the loaded_models cache to the garbage collector as early as possible. More large memory allocations happen further along in this function, so I reckon the sooner we release refs to  unneeded large objects, the better.

File: cubical/data_handler.py (hunk 1 above)
Comment (o-smirnov) [reply cubical/data_handler.py]: Done.

File: cubical/data_handler.py (hunk 2 above)
Comment (o-smirnov) [reply cubical/data_handler.py]: Done.

File: cubical/machines/abstract_machine.py (hunk 3 above)
Comment (o-smirnov) [reply cubical/machines/abstract_machine.py]: Done.

File: cubical/machines/abstract_machine.py (hunk 4 above)
Comment (o-smirnov) [reply cubical/machines/abstract_machine.py]: Done.

File: cubical/machines/abstract_machine.py (hunk 5 above)
Comment (o-smirnov) [reply cubical/machines/abstract_machine.py]: Done.

File: cubical/machines/abstract_machine.py (hunk 6 above)
Comment (o-smirnov) [reply cubical/machines/abstract_machine.py]: Done.

File: cubical/solver.py (hunk 8 above)
Comment (o-smirnov) [reply cubical/solver.py]: > ***once*** debugging is finished

Ahh, so young, so optimistic. This is exactly why this cynical old man is leaving it in. **chortle**

(Did you ever notice the Dutch have difficulty in using *if* versus *when* properly?)

Comment (JSKenyon) [reply cubical/solver.py]: Hahaha. :+1:

**Comment by JSKenyon** [comment]
Ok - I am going to merge this. For anyone using CubiCal who sees this message - the parset has changed a bit. You may need to revise existing parsets to obtain identical results.

**Comment by o-smirnov** [comment]
Haha, now that's what I call closure. This just halved the open issue count...

**Comment by o-smirnov** [comment]
We should probably do some last sanity checks, then release...

**Pull request merged**